community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/PropertyRecordCursor.java

```diff
@@ -0,0 +1,125 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.api.store;
+
+import java.io.IOException;
+
+import org.neo4j.io.pagecache.PageCursor;
+import org.neo4j.io.pagecache.PagedFile;
+import org.neo4j.kernel.impl.store.PropertyStore;
+import org.neo4j.kernel.impl.store.UnderlyingStorageException;
+import org.neo4j.kernel.impl.store.record.PropertyRecord;
+import org.neo4j.kernel.impl.store.record.Record;
+
+/**
+ * Cursor over {@link PropertyRecord} instances.
+ *
+ * Used by {@link StorePropertyCursor} to find all properties of a node or relationship.
+ *
+ */
+public class PropertyRecordCursor
+{
+    private long currentRecordId;
+    private PropertyStore propertyStore;
+
+    private PropertyRecord record;
+
+    private PageCursor pageCursor;
+
+    public PropertyRecordCursor( PropertyRecord record, PropertyStore propertyStore  )
+    {
+        this.record = record;
+        this.propertyStore = propertyStore;
+    }
+
+    public void init( long firstRecordId)
+    {
+        if (pageCursor != null)
+            throw new IllegalStateException( "Cursor already in use" );
+
+        this.currentRecordId = firstRecordId;
+    }
+
+    public boolean next()
+    {
+        try
+        {
+            if (pageCursor == null)
+            {
+                try
+                {
+                    pageCursor = propertyStore.newPageCursor( PagedFile.PF_SHARED_LOCK );
+                }
+                catch ( IOException e )
+                {
+                    throw new UnderlyingStorageException( e );
+                }
+            }
+
+            if (currentRecordId != Record.NO_NEXT_PROPERTY.intValue())
+            {
+                try
+                {
+                    pageCursor.rewind();
+                }
+                catch ( IOException e )
+                {
+                    throw new UnderlyingStorageException( e );
+                }
+                record.setId( currentRecordId );
+                propertyStore.getRecord( record, pageCursor );
+                currentRecordId = record.getNextProp();
+
+                return true;
+            } else
+            {
+                return false;
+            }
+        }
+        finally
+        {
+            // TODO We want to avoid this, but necessary for now until concurrent cursors per thread is possible
+            pageCursor.close();
```

Comment:
But pageCursor cannot be null here, right, considering the if at the top of the method. Or am I missing something?